src/main.rs

```diff
@@ -57,51 +61,39 @@ async fn activity_loop<T: process::Runtime + Clone + Unpin + 'static>(
     report_url: &str,
     activity_id: &str,
     process: ProcessController<T>,
-    agreement: AgreementDesc,
+    counters: Addr<CountersService>,
 ) -> anyhow::Result<()> {
     let report_service = gsb::service(report_url);
-    let start = Utc::now();
-    let mut current_usage = agreement.clean_usage_vector();
-    let duration_idx = agreement.resolve_counter("golem.usage.duration_sec");
 
     while let Some(()) = process.report() {
-        let now = Utc::now();
-        let duration = now - start;
-
-        if let Some(idx) = duration_idx {
-            current_usage[idx] = duration.to_std()?.as_secs_f64();
-        }
-        match report_service
-            .call(activity::local::SetUsage {
-                activity_id: activity_id.to_string(),
-                usage: ActivityUsage {
-                    current_usage: Some(current_usage.clone()),
-                    timestamp: now.timestamp(),
+        match counters.send(GetCounters).await {
+            Ok(resp) => match resp {
+                Ok(current_usage) => {
+                    set_usage_msg(&report_service, activity_id, current_usage).await
+                }
+                Err(err) => match err {
+                    CounterError::UsageLimitExceeded(info) => {
+                        set_terminate_state_msg(
+                            &report_service,
+                            activity_id,
+                            Some(format!("Usage limit exceeded: {}", info)),
+                            None,
+                        )
+                        .await
```

Comment:
Side effect of sharing `counters` implementation is an `UsageLimitExceeded` error. Not expected but handling it will do no harm.